src/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -219,6 +219,21 @@ private static (CompletionList CompletionList, bool IsIncomplete) FilterCompleti
             // Next, we sort the list based on the pattern matching result.
             matchResultsBuilder.Sort(MatchResult.SortingComparer);
 
+            // Determine if the list should be hard selected or soft selected.
+            var isFilterTextAllPunctuation = CompletionService.IsAllPunctuation(filterText);
+
+            // If we only had punctuation - we set soft selection and the list to be incomplete so we get called back when the user continues typing.
+            // If they type something that is not punctuation, we may need to update the hard vs soft selection.
+            // For example, typing '_' should initially be soft selection, but if the user types 'o' we should hard select '_otherVar' (if it exists).
+            // This isn't perfect - ideally we would make this determination every time a filter character is typed, but we do not get called back
+            // for typing filter characters in LSP (unless we always set isIncomplete, which is expensive).
+            var isHardSelection = completionList.SuggestionModeItem is null && !isFilterTextAllPunctuation;
+            var isIncomplete = isFilterTextAllPunctuation;
+
+            // If our completion list hasn't hit the max size, we don't need to do anything filtering
+            if (completionListMaxSize < 0 || completionListMaxSize >= completionList.ItemsList.Count)
+                return (completionList, isIncomplete, isHardSelection);
```

Comment:
I think this part can be moved to the beginning of the method, before calculating MatchResults

Reply:
ah yup now that it no longer needs it.  moved